File: 0142-linked-list-cycle-ii/solution.py
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        # (base case)
        if not head or not head.next: return None
        if head == head.next.next: return head

        # ==================================================
        #  Linked List + Two Pointer (Floyd's algorithm)   =
        # ==================================================
        # time  : O(n)
        # space : O(1)

        slowP, fastP = head, head

        # Find the intersection
        while fastP and fastP.next:
            slowP = slowP.next
            fastP = fastP.next.next
            if slowP == fastP: break

        if slowP != fastP: return None

        # Find the cycle entrance
        slowP = head
        while slowP != fastP:
            slowP = slowP.next
            fastP = fastP.next

        return slowP

File: 0142-linked-list-cycle-ii/solution.java
class Solution {
    /**
     * @time  : O(n)
     * @space : O(1)
     */
    
    public ListNode detectCycle(ListNode head) {
        /* case case */
        if(head == null || head.next == null) return null;
        if(head == head.next.next) return head;
        
        ListNode slowP = head, fastP = head;
        while(fastP != null && fastP.next != null) {
            slowP = slowP.next;
            fastP = fastP.next.next;
            if(slowP == fastP) break;
        }
        
        if(slowP != fastP) return null;
        
        slowP = head;
        while(slowP != fastP) {
            slowP = slowP.next;
            fastP = fastP.next;
        }
        
        return slowP;
    }
}
